### PR TITLE
Using pyo3's signature information to generate python's defualt value

### PR DESCRIPTION
**Pull request by zao111222333**
Hi everyone,
I want to improve the function signature when pyo3's signature is specified. 
Let me use an example to show what did I do, for the following rust function:

```rust
/// `result=(a * b) + c`,
/// `b` and `c` has defualt values
#[gen_stub_pyfunction]
#[pyfunction]
#[pyo3(signature = (a, b = vec![1,2], c = 2 ))]
fn mul_add_vec(a: usize, b: Vec<usize>, c: usize) -> Vec<usize> {
    b.into_iter().map(|b| a * b + c).collect()
}
```
Original output
```python
def mul_add_vec(a: ..., b: ... = ..., c: ... = ...) -> list[int]:
    ...
```
Now
```python
def mul_add_vec(a:int, b:typing.Sequence[int]=[1, 2], c:int=2) -> list[int]:
    ...
```
## My approach:
For instance, how to infer python's `[1, 2]` from rust's `vec![1,2]`?
To summary, I use `pyo3::IntoPyObject::into_pyobject` to get the python object converted from `vec![1,2]`, then call `to_string` to get the python object's value.

More deltails:
+ The convertion is now static, so I use `std::sync::LazyLock`, see [here](
https://github.com/zao111222333/pyo3-stub-gen/blob/56f7af96985c6909265f13766844363710da663f/pyo3-stub-gen-derive/src/gen_stub/signature.rs#L103-L113)
+ In `pyo3-stub-gen`, I merge the signature information into each `ArgInfo`, see [here](https://github.com/zao111222333/pyo3-stub-gen/blob/56f7af96985c6909265f13766844363710da663f/pyo3-stub-gen/src/generate/arg.rs#L8)

### CODE REVIEW

File: pyo3-stub-gen-derive/src/gen_stub/member.rs [hunk 1]
```diff
@@ -29,12 +30,17 @@ impl TryFrom<ImplItemFn> for MemberInfo {
...
         assert!(Self::is_candidate_item(&item)?);
         let ImplItemFn { attrs, sig, .. } = &item;
-        let attrs = parse_pyo3_attrs(attrs)?;
-        for attr in attrs {
+        for attr in parse_pyo3_attrs(attrs)? {
             if let Attr::Getter(name) = attr {
+                let name = name.unwrap_or(sig.ident.to_string());
                 return Ok(MemberInfo {
-                    name: name.unwrap_or(sig.ident.to_string()),
+                    name: if let Some(name) = name.strip_prefix("get_") {
```
Comment (konn) [inline pyo3-stub-gen-derive/src/gen_stub/member.rs]: How does this change related to the topic of this PR?

File: pyo3-stub-gen-derive/src/gen_stub/pyclass.rs [hunk 2]
```diff
@@ -152,4 +157,4 @@ mod test {
             .unwrap()
             .to_string()
     }
-}
```
Comment (konn) [inline pyo3-stub-gen-derive/src/gen_stub/pyclass.rs]: Please add extra newline here.

File: pyo3-stub-gen/src/generate/member.rs [hunk 3]
```diff
@@ -19,13 +20,22 @@ impl From<&MemberInfo> for MemberDef {
...
+            writeln!(f, r#"{indent}r""""#)?;
+            for line in self.doc.lines() {
+                writeln!(f, "{indent}{}", line)?;
+            }
+            writeln!(f, r#"{indent}""""#)?;
+        }
+        Ok(())
     }
-}
+}
```
Comment (konn) [inline pyo3-stub-gen/src/generate/member.rs]: Please add a newline here, too.

File: pyo3-stub-gen-derive/src/gen_stub/signature.rs [hunk 4]
```diff
@@ -70,15 +62,108 @@ impl Parse for Signature {
...
+                            r#type: <#ty as ::pyo3_stub_gen::PyStubType>::type_input,
+                            signature: Some(pyo3_stub_gen::type_info::SignatureArg::Assign{
+                                default: {
+                                    static DEFAULT: std::sync::LazyLock<String> = std::sync::LazyLock::new(|| {
+                                        pyo3::prepare_freethreaded_python();
+                                        ::pyo3::Python::with_gil(|py| -> String {
+                                            let v: #ty = #value;
+                                            <#ty as ::pyo3::IntoPyObject>::into_pyobject(v, py)
+                                                .unwrap()
+                                                .to_string()
```
Comment (konn) [inline pyo3-stub-gen-derive/src/gen_stub/signature.rs]: Unfortunately, `to_string()`-based approach works only for values with concrete python values, but not for other types like pointes to Rust data.

For example, we can call _any_ function returning `PyFunctionArg` types in default argument position. So, the following code compiles:

```rust
#[gen_stub_pyclass]
#[pyclass]
#[derive(Debug)]
struct A {
    #[pyo3(get, set)]
    x: usize,
}

fn mkpyA() -> Py<A> {
    Python::with_gil(|py| Py::new(py, A { x: 12 }).unwrap())
}

#[gen_stub_pyfunction]
#[pyfunction]
#[pyo3(signature = (a = mkpyA()))]
fn default_with_a(py: Python, a: Py<A>) -> String {
    format!("{:?}", a.borrow(py))
}
```

The above code code generates the following stub with this PR:

```pyi
def default_with_a(a:A=<builtins.A object at 0x105a86970>) -> str:
    ...
```

Note that `<builtins.A object at 0x105a86970>` is not a valid python expression and pyright complains about it.

Hence, we cannot use `to_string()` here - indeed, some types in our production codebase has `to_string()` implementation invalid as a python expression. Hence we cannot accept this as-is.

To achieve your goal in backward-compatible manner, there are several options:

1. use `eval` to parse `to_string()` result back to the original type and see if they coincide and use `...` as a fallback
  - Pros: any types with concrete and faithful representation can have default value.
  - Cons: This can allow arbitrary code execution at the compile time (say, via Pull Request CI)
2. Try to `extract()` pre-specified primitive types (e.g. `None`, numbers, strings, and booleanas) from `v.into_pyobject(py)`, and use `to_string()` for them only; otherwise use `...` instead.
  - Pros: This doesn't seem to cause any security incident.
  - Cons: This is rather ad-hoc.
 
(Other possible way is to introduce some trait(s) for textual default option rendering, but it breaks backward-compatibility.)
As option (1) has a security concern, please rewrite this with option (2). To support broader range of values, perhaps you may want to use [`serder-pyobject`](https://crates.io/crates/serde-pyobject) crate, which is also developed by us.

**Review by konn — CHANGES_REQUESTED** [review]
Hi, thank you for your contribution! We are surprised that we are just planning to work on coexistence of signature and type hints in function arguments, and, alas, your PR include this feature! What a great synchronicity. 😄 

We are willing to merge your PR, provided that corner cases in `to_string()`-based approach are fixed; please see the review below.

**Review by konn — COMMENTED** [review]
It seems that some CI are failing. Please apply rustfmt and clippy accordingly.

Also, if you don't have enough time, I can do the rest of the work, either by creating another PR onto your branch, or directly edit your branch. Please feel free to call for help. :+1:

**Comment by zao111222333** [comment]
Hi @konn, thank you so much for your review & reply!
Yes, I will try to use `eval` or `extract` to fix the corner cases. 
For getter's change, I update some lines to propagate getter's `doc`, it's not releative to this PR, I will remove them and create a new PR for that feature.

**Comment by konn** [comment]
> Hi @konn, thank you so much for your review & reply! Yes, I will try to use `eval` or `extract` to fix the corner cases.

Thanks! As pointed out above, please avoid using `eval` as it has some security concern.

> For getter's change, I update some lines to propagate getter's `doc`, it's not releative to this PR, I will remove them and create a new PR for that feature.

:+1:

**Comment by zao111222333** [comment]
Hi @konn , I replace `to_string` with `extract`, it's a little messy but verified works in my [test cases](
https://github.com/zao111222333/pyo3-stub-gen/blob/d0ae53439a63fe171688021820ce39f134ba0696/pyo3-stub-gen/src/util.rs#L61-L126
).
Waitting for your feedback :)

File: pyo3-stub-gen/src/util.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
+use pyo3::{prelude::*, types::*};
+
+/// Reference to https://github.com/Jij-Inc/serde-pyobject/blob/5916fefc5b7f0e096ed13869dc41b36ad6f62dc2/src/de.rs#L309-L333
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: This comment doesn't have to be doc comment (`///`); perhaps you can just drop this.

File: pyo3-stub-gen/src/util.rs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+pub fn fmt_py_obj(any: &Bound<'_, PyAny>) -> Result<String, ()> {
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: As [clippy pointed out](https://github.com/Jij-Inc/pyo3-stub-gen/actions/runs/12784706370/job/35642005732?pr=135), using `()` as a Error type is not recommended. I think it is enough to use `Option` instead, or reuse `pyo3`'s `Result` synonym and returning `PyValueError::new(...)` would suffice.

File: pyo3-stub-gen/src/util.rs [hunk 7]
```diff
@@ -0,0 +1,145 @@
+use pyo3::{prelude::*, types::*};
+
+/// Reference to <https://github.com/Jij-Inc/serde-pyobject/blob/5916fefc5b7f0e096ed13869dc41b36ad6f62dc2/src/de.rs#L309-L333>
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: Perhaps we don't need this reference URL (at least as a doc comment (e.g. comments starting with `/// `)), as the intention of the code is clear enough.

File: pyo3-stub-gen/src/util.rs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+#[allow(clippy::result_unit_err)]
+pub fn fmt_py_obj(any: &Bound<'_, PyAny>) -> Result<String, ()> {
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: Also, I think formatting `dict`, `list`, and `tuple`s manually can contain some corner-cases. Rather, I recommend to check if the values are consisting only of python values, and then use `to_string()` once the invariance is checked. See below for example.

File: pyo3-stub-gen/src/util.rs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                .collect::<Result<Vec<String>, ()>>()
+                .map(|v_list| format!("[{}]", v_list.join(", ")))
+        });
+    }
+    if any.is_instance_of::<PyTuple>() {
+        return any.downcast::<PyTuple>().map_err(|_| ()).and_then(|list| {
+            list.into_iter()
+                .map(|v| fmt_py_obj(&v))
+                .collect::<Result<Vec<String>, ()>>()
+                .map(|v_list| format!("({})", v_list.join(", ")))
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: This is one manifestation of corner-cases of manual formatting. I think it formats the tuple `(42,)`  (i.e. a singleton tuple with `42` its only component) as `(42)`, which will be interpreted as a mere number literal `42`.

**Review by konn — CHANGES_REQUESTED** [review]
Thank you for your swift reaction and detailed tests! I think we are almost done, but please check some additional reviews.

**Comment by konn** [comment]
> Sorry for forgetting the CI check for util.rs, I fixed them. However, I'm not so sure how to fix the errors from `cargo semver-checks`.

Thanks! I've just edited your branch so that `semver-checks` CI would success. Please pull before working on my change.

**Comment by zao111222333** [comment]
> Also, I think formatting `dict`, `list`, and `tuple`s manually can contain some corner-cases. Rather, I recommend to check if the values are consisting only of python values, and then use `to_string()` once the invariance is checked. See below for example.

That's a great idea and we can have a more elegant formatter function. I finished that.

File: pyo3-stub-gen/src/util.rs [hunk 10]
```diff
@@ -0,0 +1,114 @@
+use pyo3::{prelude::*, types::*};
+
+pub fn fmt_py_obj(any: &Bound<'_, PyAny>) -> String {
+    fn all_builtin_types(any: &Bound<'_, PyAny>) -> bool {
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: Perhaps we can make `all_builtin_types` as independent function for future use?

File: pyo3-stub-gen/src/util.rs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        // `to_string` for `PyString` will lose quotes
+        if any.is_instance_of::<PyString>() {
+            if let Ok(s) = any.extract::<String>() {
+                return format!("'{s}'");
+            }
+        } else {
+            return any.to_string();
+        }
+    }
+    "...".to_owned()
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs]: Ah, I am not aware this behaviour. And this formatting seems generate invalid output when it is fed wiith string like `"foo'bar"` or `"buz\\"`.
It seems that python provides [`repr` function to convert into evalable types (for builtin values)](https://docs.python.org/3.10/library/functions.html#repr) and PyO3 provides `Bound::repr` counterpart. So I think this is enough:

```suggestion
    if all_builtin_types(any) {
        any.repr().map_or("...".to_string(), |a| a.to_string())
    } else {
        "...".to_string()
    }
```

File: pyo3-stub-gen/src/util.rs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            let tuple = PyTuple::new(py, [A {}]).unwrap();
+            assert_eq!("...", fmt_py_obj(&tuple));
+        })
+    }
+    #[test]
+    fn test_fmt_other() {
+        pyo3::prepare_freethreaded_python();
+        Python::with_gil(|py| {
+            // str
+            assert_eq!("'123'", fmt_py_obj(&"123".into_bound_py_any(py).unwrap()));
```
Comment (konn) [inline pyo3-stub-gen/src/util.rs:95]: Let's test corner cases for string literals as well.

```suggestion
            assert_eq!("'123'", fmt_py_obj(&"123".into_bound_py_any(py).unwrap()));
            assert_eq!(
                "\"don't\"",
                fmt_py_obj(&"don't".into_bound_py_any(py).unwrap())
            );
            assert_eq!(
                "'str\\\\'",
                fmt_py_obj(&"str\\".into_bound_py_any(py).unwrap())
            );
```

**Review by konn — CHANGES_REQUESTED** [review]
Thank you for your change! I am not aware that `to_string()` cannot be used for string literals. It seems calling `repr()` is more suitable for this kind of conversion and does string escape properly. See suggestions below.

**Comment by zao111222333** [comment]
The suggestions are adopted, waitting for your feedback.
And if you are willing, I want to create a PR to do the following things:
+ Add `indexmap` feature to support `IndexMap` and `IndexSet`
+ Propagate `doc` to getter's class member
+ User specified default value for getter's class member (need `stub_gen` macro attribute), e.g.,
``` rust
#[pyclass]
#[stub_gen_pyclass]
struct A {
    #[pyo3(get,set)]
    #[stub_gen(default = 2)]
    x: usize,
}
impl Default for A {
    fn default() -> A {
        Self { x: 2 }
    }
}
```
Output:
``` python
class A:
    x: int = 2
```

**Review by konn — APPROVED** [review]
LGTM :+1:, thank you for your contributions!

**Comment by konn** [comment]
And thank you for your future contributions.

> * Add `indexmap` feature to support `IndexMap` and `IndexSet`

I think you are fine to work on this PR, but there is no particular need for us for time being, it can take some time to be get merged.

> * Propagate `doc` to getter's class member

It seems duplicate of #96 - if it is different, please let us know.

> * User specified default value for getter's class member (need `stub_gen` macro attribute), e.g.,

This seems good to have, but we have to carefully discuss the design of new type of attribute macro.

Anyway, you can feel free to open PRs if it is not a duplicate of an existing ones. Thank you again for your contribution!

**Comment by zao111222333** [comment]
Thanks so much for your assistance! I will open PRs once I finish those idea.
